In [31]:
import pandas as pd

In [32]:
data = pd.read_pickle('data/03_top_liquid_stocks.pkl')

In [33]:
data

close        high  garman_klass_vol        rsi  \
date       ticker                                                        
2020-08-31 AAPL    125.057526  126.957042          0.000707  76.744809   
           ABBV     75.952614   76.142956          0.000105  52.776389   
           ABT      97.865204   99.894561          0.000204  67.784280   
           ACN     217.961929  221.241397          0.000094  67.876178   
           ADBE    513.390015  516.729980          0.000153  66.121887   
...                       ...         ...               ...        ...   
2026-08-31 WDC     454.100006  467.799988          0.000662  43.796277   
           WFC      88.930000   89.220001          0.000170  59.174628   
           WMT     116.010002  116.050003          0.000153  54.108979   
           XOM     158.722244  159.199152          0.000119  62.142821   
           XYZ      78.279999   79.199997          0.000312  49.935831   

                     bb_low    bb_mid   bb_high       atr      macd  \
date       ticker                                                     
2020-08-31 AAPL    4.634849  4.743995  4.853141 -0.215957  1.968695   
           ABBV    4.304350  4.329241  4.354133 -1.060942 -0.352670   
           ABT     4.460974  4.528430  4.595886  0.635079  1.529678   
           ACN     5.324580  5.364798  5.405016 -1.217038  0.870915   
           ADBE    6.036542  6.152162  6.267782  0.679071  1.881340   
...                     ...       ...       ...       ...       ...   
2026-08-31 WDC     6.028525  6.205588  6.382652  5.179602 -4.059220   
           WFC     4.444806  4.474222  4.503639  1.216887  0.744505   
           WMT     4.693527  4.727599  4.761672  2.212910 -0.599633   
           XOM     4.988154  5.038804  5.089454  2.227161  1.943697   
           XYZ     4.342263  4.398908  4.455553 -0.668036  0.144249   

                   return_1m  return_2m  return_3m  return_6m  return_9m  \
date       ticker                                                          
2020-08-31 AAPL     0.204976   0.176611   0.158586   0.110026   0.076700   
           ABBV     0.009061  -0.006341   0.015125   0.023252   0.014328   
           ABT      0.087738   0.096338   0.050042   0.059706   0.029314   
           ACN      0.067399   0.059010   0.060994   0.049919   0.021139   
           ADBE     0.155451   0.085986   0.099162   0.068430   0.057830   
...                      ...        ...        ...        ...        ...   
2026-08-31 WDC     -0.166544  -0.156820  -0.050854   0.084251   0.281101   
           WFC      0.034593   0.040332   0.048744   0.016685   0.005747   
           WMT      0.043255   0.012067   0.000748  -0.015546   0.006081   
           XOM      0.027728   0.080947   0.032203   0.008908   0.037822   
           XYZ     -0.036435   0.014889   0.011145   0.034948   0.017777   

                   return_12m  
date       ticker              
2020-08-31 AAPL      0.073806  
           ABBV      0.036576  
           ABT       0.022377  
           ACN       0.017444  
           ADBE      0.050419  
...                       ...  
2026-08-31 WDC       0.195755  
           WFC       0.008411  
           WMT       0.015550  
           XOM       0.030196  
           XYZ      -0.001434  

[10419 rows x 15 columns]

In [34]:
import pandas_datareader.data as web

In [35]:
factor_data = web.DataReader('F-F_Research_Data_5_Factors_2x3',
                                'famafrench',
                                start='2010')[0].drop('RF', axis=1)

In [36]:
factor_data

,Mkt-RF,SMB,HML,RMW,CMA
Date,,,,,
2010-01,-3.35,0.40,0.33,-1.08,0.51
2010-02,3.39,1.49,3.18,-0.29,1.42
2010-03,6.30,1.83,2.19,-0.61,1.74
2010-04,2.00,4.96,2.96,0.61,1.75
2010-05,-7.90,0.08,-2.48,1.30,-0.24
...,...,...,...,...,...
2026-02,-1.18,0.74,2.71,1.91,5.04
2026-03,-5.18,0.66,3.41,-2.08,-0.10
2026-04,9.95,0.38,-1.17,-4.30,-3.78


In [37]:
factor_data.index = factor_data.index.to_timestamp()

In [38]:
factor_data

,Mkt-RF,SMB,HML,RMW,CMA
Date,,,,,
2010-01-01,-3.35,0.40,0.33,-1.08,0.51
2010-02-01,3.39,1.49,3.18,-0.29,1.42
2010-03-01,6.30,1.83,2.19,-0.61,1.74
2010-04-01,2.00,4.96,2.96,0.61,1.75
2010-05-01,-7.90,0.08,-2.48,1.30,-0.24
...,...,...,...,...,...
2026-02-01,-1.18,0.74,2.71,1.91,5.04
2026-03-01,-5.18,0.66,3.41,-2.08,-0.10
2026-04-01,9.95,0.38,-1.17,-4.30,-3.78


In [39]:
factor_data.index.name = 'date'
factor_data

,Mkt-RF,SMB,HML,RMW,CMA
date,,,,,
2010-01-01,-3.35,0.40,0.33,-1.08,0.51
2010-02-01,3.39,1.49,3.18,-0.29,1.42
2010-03-01,6.30,1.83,2.19,-0.61,1.74
2010-04-01,2.00,4.96,2.96,0.61,1.75
2010-05-01,-7.90,0.08,-2.48,1.30,-0.24
...,...,...,...,...,...
2026-02-01,-1.18,0.74,2.71,1.91,5.04
2026-03-01,-5.18,0.66,3.41,-2.08,-0.10
2026-04-01,9.95,0.38,-1.17,-4.30,-3.78


In [40]:
data

close        high  garman_klass_vol        rsi  \
date       ticker                                                        
2020-08-31 AAPL    125.057526  126.957042          0.000707  76.744809   
           ABBV     75.952614   76.142956          0.000105  52.776389   
           ABT      97.865204   99.894561          0.000204  67.784280   
           ACN     217.961929  221.241397          0.000094  67.876178   
           ADBE    513.390015  516.729980          0.000153  66.121887   
...                       ...         ...               ...        ...   
2026-08-31 WDC     454.100006  467.799988          0.000662  43.796277   
           WFC      88.930000   89.220001          0.000170  59.174628   
           WMT     116.010002  116.050003          0.000153  54.108979   
           XOM     158.722244  159.199152          0.000119  62.142821   
           XYZ      78.279999   79.199997          0.000312  49.935831   

                     bb_low    bb_mid   bb_high       atr      macd  \
date       ticker                                                     
2020-08-31 AAPL    4.634849  4.743995  4.853141 -0.215957  1.968695   
           ABBV    4.304350  4.329241  4.354133 -1.060942 -0.352670   
           ABT     4.460974  4.528430  4.595886  0.635079  1.529678   
           ACN     5.324580  5.364798  5.405016 -1.217038  0.870915   
           ADBE    6.036542  6.152162  6.267782  0.679071  1.881340   
...                     ...       ...       ...       ...       ...   
2026-08-31 WDC     6.028525  6.205588  6.382652  5.179602 -4.059220   
           WFC     4.444806  4.474222  4.503639  1.216887  0.744505   
           WMT     4.693527  4.727599  4.761672  2.212910 -0.599633   
           XOM     4.988154  5.038804  5.089454  2.227161  1.943697   
           XYZ     4.342263  4.398908  4.455553 -0.668036  0.144249   

                   return_1m  return_2m  return_3m  return_6m  return_9m  \
date       ticker                                                          
2020-08-31 AAPL     0.204976   0.176611   0.158586   0.110026   0.076700   
           ABBV     0.009061  -0.006341   0.015125   0.023252   0.014328   
           ABT      0.087738   0.096338   0.050042   0.059706   0.029314   
           ACN      0.067399   0.059010   0.060994   0.049919   0.021139   
           ADBE     0.155451   0.085986   0.099162   0.068430   0.057830   
...                      ...        ...        ...        ...        ...   
2026-08-31 WDC     -0.166544  -0.156820  -0.050854   0.084251   0.281101   
           WFC      0.034593   0.040332   0.048744   0.016685   0.005747   
           WMT      0.043255   0.012067   0.000748  -0.015546   0.006081   
           XOM      0.027728   0.080947   0.032203   0.008908   0.037822   
           XYZ     -0.036435   0.014889   0.011145   0.034948   0.017777   

                   return_12m  
date       ticker              
2020-08-31 AAPL      0.073806  
           ABBV      0.036576  
           ABT       0.022377  
           ACN       0.017444  
           ADBE      0.050419  
...                       ...  
2026-08-31 WDC       0.195755  
           WFC       0.008411  
           WMT       0.015550  
           XOM       0.030196  
           XYZ      -0.001434  

[10419 rows x 15 columns]

In [41]:
factor_data.index


DatetimeIndex(['2010-01-01', '2010-02-01', '2010-03-01', '2010-04-01',
               '2010-05-01', '2010-06-01', '2010-07-01', '2010-08-01',
               '2010-09-01', '2010-10-01',
               ...
               '2025-09-01', '2025-10-01', '2025-11-01', '2025-12-01',
               '2026-01-01', '2026-02-01', '2026-03-01', '2026-04-01',
               '2026-05-01', '2026-06-01'],
              dtype='datetime64[us]', name='date', length=198, freq='MS')

In [42]:
factor_data

,Mkt-RF,SMB,HML,RMW,CMA
date,,,,,
2010-01-01,-3.35,0.40,0.33,-1.08,0.51
2010-02-01,3.39,1.49,3.18,-0.29,1.42
2010-03-01,6.30,1.83,2.19,-0.61,1.74
2010-04-01,2.00,4.96,2.96,0.61,1.75
2010-05-01,-7.90,0.08,-2.48,1.30,-0.24
...,...,...,...,...,...
2026-02-01,-1.18,0.74,2.71,1.91,5.04
2026-03-01,-5.18,0.66,3.41,-2.08,-0.10
2026-04-01,9.95,0.38,-1.17,-4.30,-3.78


In [43]:
factor_data.resample('ME').last().div(100)

,Mkt-RF,SMB,HML,RMW,CMA
date,,,,,
2010-01-31,-0.0335,0.0040,0.0033,-0.0108,0.0051
2010-02-28,0.0339,0.0149,0.0318,-0.0029,0.0142
2010-03-31,0.0630,0.0183,0.0219,-0.0061,0.0174
2010-04-30,0.0200,0.0496,0.0296,0.0061,0.0175
2010-05-31,-0.0790,0.0008,-0.0248,0.0130,-0.0024
...,...,...,...,...,...
2026-02-28,-0.0118,0.0074,0.0271,0.0191,0.0504
2026-03-31,-0.0518,0.0066,0.0341,-0.0208,-0.0010
2026-04-30,0.0995,0.0038,-0.0117,-0.0430,-0.0378


In [44]:
factor_data = factor_data.resample('ME').last().div(100)

In [45]:
factor_data

,Mkt-RF,SMB,HML,RMW,CMA
date,,,,,
2010-01-31,-0.0335,0.0040,0.0033,-0.0108,0.0051
2010-02-28,0.0339,0.0149,0.0318,-0.0029,0.0142
2010-03-31,0.0630,0.0183,0.0219,-0.0061,0.0174
2010-04-30,0.0200,0.0496,0.0296,0.0061,0.0175
2010-05-31,-0.0790,0.0008,-0.0248,0.0130,-0.0024
...,...,...,...,...,...
2026-02-28,-0.0118,0.0074,0.0271,0.0191,0.0504
2026-03-31,-0.0518,0.0066,0.0341,-0.0208,-0.0010
2026-04-30,0.0995,0.0038,-0.0117,-0.0430,-0.0378


In [46]:
factor_data = factor_data.join(data['return_1m']).sort_index()

In [47]:
factor_data


Mkt-RF     SMB     HML     RMW     CMA  return_1m
date       ticker                                                   
2020-08-31 AAPL    0.0762 -0.0089 -0.0302  0.0427 -0.0151   0.204976
           ABBV    0.0762 -0.0089 -0.0302  0.0427 -0.0151   0.009061
           ABT     0.0762 -0.0089 -0.0302  0.0427 -0.0151   0.087738
           ACN     0.0762 -0.0089 -0.0302  0.0427 -0.0151   0.067399
           ADBE    0.0762 -0.0089 -0.0302  0.0427 -0.0151   0.155451
...                   ...     ...     ...     ...     ...        ...
2026-06-30 WDC    -0.0107  0.0458  0.0334 -0.0464  0.0346   0.202700
           WFC    -0.0107  0.0458  0.0334 -0.0464  0.0346   0.065772
           WMT    -0.0107  0.0458  0.0334 -0.0464  0.0346  -0.021512
           XOM    -0.0107  0.0458  0.0334 -0.0464  0.0346  -0.058791
           XYZ    -0.0107  0.0458  0.0334 -0.0464  0.0346   0.003698

[10137 rows x 6 columns]

In [48]:
factor_data.xs('AAPL',level=1)

,Mkt-RF,SMB,HML,RMW,CMA,return_1m
date,,,,,,
2020-08-31,0.0762,-0.0089,-0.0302,0.0427,-0.0151,0.204976
2020-09-30,-0.0364,-0.0001,-0.0272,-0.0138,-0.0193,-0.102526
2020-10-31,-0.0208,0.0467,0.0425,-0.0077,-0.0057,-0.060012
2020-11-30,0.1244,0.0712,0.0209,-0.0225,0.0128,0.095493
2020-12-31,0.0463,0.0493,-0.0168,-0.0197,-0.0006,0.114574
...,...,...,...,...,...,...
2026-02-28,-0.0118,0.0074,0.0271,0.0191,0.0504,0.019066
2026-03-31,-0.0518,0.0066,0.0341,-0.0208,-0.0010,-0.039329
2026-04-30,0.0995,0.0038,-0.0117,-0.0430,-0.0378,0.069191


In [50]:
observations = factor_data.groupby(level=1).size()

In [55]:
observations

ticker
AAPL    71
ABBV    71
ABNB    43
ABT     71
ACN     71
        ..
WFC     71
WMT     71
WYNN    32
XOM     71
XYZ     71
Length: 174, dtype: int64

In [51]:
valid_stocks = observations[observations >= 10]

In [52]:
valid_stocks

ticker
AAPL    71
ABBV    71
ABNB    43
ABT     71
ACN     71
        ..
WFC     71
WMT     71
WYNN    32
XOM     71
XYZ     71
Length: 163, dtype: int64

In [53]:
factor_data.index.get_level_values('ticker').unique()

Index(['AAPL', 'ABBV', 'ABT', 'ACN', 'ADBE', 'ADP', 'ADSK', 'ALGN', 'AMAT',
       'AMD',
       ...
       'DELL', 'HOOD', 'CEG', 'ETN', 'APP', 'GEV', 'RDDT', 'SNPS', 'COF',
       'PGR'],
      dtype='str', name='ticker', length=174)

In [54]:
valid_stocks.index.get_level_values('ticker').unique()

Index(['AAPL', 'ABBV', 'ABNB', 'ABT', 'ACN', 'ADBE', 'ADI', 'ADP', 'ADSK',
       'ALGN',
       ...
       'VLO', 'VRTX', 'VZ', 'WDAY', 'WDC', 'WFC', 'WMT', 'WYNN', 'XOM', 'XYZ'],
      dtype='str', name='ticker', length=163)

In [59]:
factor_data.index.get_level_values('ticker')

Index(['AAPL', 'ABBV', 'ABT', 'ACN', 'ADBE', 'ADP', 'ADSK', 'ALGN', 'AMAT',
       'AMD',
       ...
       'UPS', 'V', 'VRTX', 'VZ', 'WDAY', 'WDC', 'WFC', 'WMT', 'XOM', 'XYZ'],
      dtype='str', name='ticker', length=10137)

In [64]:
factor_data.index

MultiIndex([('2020-08-31', 'AAPL'),
            ('2020-08-31', 'ABBV'),
            ('2020-08-31',  'ABT'),
            ('2020-08-31',  'ACN'),
            ('2020-08-31', 'ADBE'),
            ('2020-08-31',  'ADP'),
            ('2020-08-31', 'ADSK'),
            ('2020-08-31', 'ALGN'),
            ('2020-08-31', 'AMAT'),
            ('2020-08-31',  'AMD'),
            ...
            ('2026-06-30',  'UPS'),
            ('2026-06-30',    'V'),
            ('2026-06-30', 'VRTX'),
            ('2026-06-30',   'VZ'),
            ('2026-06-30', 'WDAY'),
            ('2026-06-30',  'WDC'),
            ('2026-06-30',  'WFC'),
            ('2026-06-30',  'WMT'),
            ('2026-06-30',  'XOM'),
            ('2026-06-30',  'XYZ')],
           names=['date', 'ticker'], length=10137)

In [61]:
factor_data.index.get_level_values('date')

DatetimeIndex(['2020-08-31', '2020-08-31', '2020-08-31', '2020-08-31',
               '2020-08-31', '2020-08-31', '2020-08-31', '2020-08-31',
               '2020-08-31', '2020-08-31',
               ...
               '2026-06-30', '2026-06-30', '2026-06-30', '2026-06-30',
               '2026-06-30', '2026-06-30', '2026-06-30', '2026-06-30',
               '2026-06-30', '2026-06-30'],
              dtype='datetime64[us]', name='date', length=10137, freq=None)

In [62]:
factor_data.index.get_level_values('ticker').isin(valid_stocks.index)

array([ True,  True,  True, ...,  True,  True,  True], shape=(10137,))

In [70]:
factor_data = factor_data[factor_data.index.get_level_values('ticker').isin(valid_stocks.index)]

Now we have only those stocks which appeared for more than 10 months , so we will be training our model on regular stocks

10137 -10093 = 44 rows got trimmed down because they had stocks which had less than 10 months of data


Now lets attach the weights to our features using Rolling OLS

In [82]:
factor_data


Mkt-RF     SMB     HML     RMW     CMA  return_1m
date       ticker                                                   
2020-08-31 AAPL    0.0762 -0.0089 -0.0302  0.0427 -0.0151   0.204976
           ABBV    0.0762 -0.0089 -0.0302  0.0427 -0.0151   0.009061
           ABT     0.0762 -0.0089 -0.0302  0.0427 -0.0151   0.087738
           ACN     0.0762 -0.0089 -0.0302  0.0427 -0.0151   0.067399
           ADBE    0.0762 -0.0089 -0.0302  0.0427 -0.0151   0.155451
...                   ...     ...     ...     ...     ...        ...
2026-06-30 WDC    -0.0107  0.0458  0.0334 -0.0464  0.0346   0.202700
           WFC    -0.0107  0.0458  0.0334 -0.0464  0.0346   0.065772
           WMT    -0.0107  0.0458  0.0334 -0.0464  0.0346  -0.021512
           XOM    -0.0107  0.0458  0.0334 -0.0464  0.0346  -0.058791
           XYZ    -0.0107  0.0458  0.0334 -0.0464  0.0346   0.003698

[10093 rows x 6 columns]

In [68]:
from statsmodels.regression.rolling import RollingOLS
import statsmodels.api as sm

In [76]:
factor_data.groupby(level=1,group_keys=False).apply(lambda x: RollingOLS(endog=x['return_1m'],
                                                                         exog=sm.add_constant(x.drop('return_1m',axis=1)),
                                                                        window=min(24,x.shape[0]),
                                                                        min_nobs=len(x.columns)+1)
   .fit(params_only=True) 
   .params.drop('const',axis=1))

Mkt-RF       SMB       HML       RMW        CMA
date       ticker                                                   
2020-08-31 AAPL         NaN       NaN       NaN       NaN        NaN
           ABBV         NaN       NaN       NaN       NaN        NaN
           ABT          NaN       NaN       NaN       NaN        NaN
           ACN          NaN       NaN       NaN       NaN        NaN
           ADBE         NaN       NaN       NaN       NaN        NaN
...                     ...       ...       ...       ...        ...
2026-06-30 WDC     1.765352 -5.613121 -2.996401 -0.411387  22.777256
           WFC     0.800225  0.185045  0.546113 -0.383377  -0.290316
           WMT     0.840213  0.133426  0.046253  1.272456  -0.500936
           XOM     0.034127 -0.063205  1.141849  0.543015   0.186066
           XYZ     0.799496  1.790561 -2.484338 -0.561827   0.130112

[10093 rows x 5 columns]

In [77]:
betas = (factor_data.groupby(level=1,group_keys=False).apply(lambda x: RollingOLS(endog=x['return_1m'],
                                                                         exog=sm.add_constant(x.drop('return_1m',axis=1)),
                                                                        window=min(24,x.shape[0]),
                                                                        min_nobs=len(x.columns)+1)
.fit(params_only=True)
.params
.drop('const',axis=1)))

betas

Mkt-RF       SMB       HML       RMW        CMA
date       ticker                                                   
2020-08-31 AAPL         NaN       NaN       NaN       NaN        NaN
           ABBV         NaN       NaN       NaN       NaN        NaN
           ABT          NaN       NaN       NaN       NaN        NaN
           ACN          NaN       NaN       NaN       NaN        NaN
           ADBE         NaN       NaN       NaN       NaN        NaN
...                     ...       ...       ...       ...        ...
2026-06-30 WDC     1.765352 -5.613121 -2.996401 -0.411387  22.777256
           WFC     0.800225  0.185045  0.546113 -0.383377  -0.290316
           WMT     0.840213  0.133426  0.046253  1.272456  -0.500936
           XOM     0.034127 -0.063205  1.141849  0.543015   0.186066
           XYZ     0.799496  1.790561 -2.484338 -0.561827   0.130112

[10093 rows x 5 columns]

In [78]:
betas.groupby('ticker').shift()

Mkt-RF       SMB       HML       RMW       CMA
date       ticker                                                  
2020-08-31 AAPL         NaN       NaN       NaN       NaN       NaN
           ABBV         NaN       NaN       NaN       NaN       NaN
           ABT          NaN       NaN       NaN       NaN       NaN
           ACN          NaN       NaN       NaN       NaN       NaN
           ADBE         NaN       NaN       NaN       NaN       NaN
...                     ...       ...       ...       ...       ...
2026-06-30 WDC          NaN       NaN       NaN       NaN       NaN
           WFC     0.925292  0.038777  0.671017 -0.121967 -0.566944
           WMT     0.897085  0.075673  0.146779  1.378219 -0.643397
           XOM    -0.243429  0.264830  0.881779 -0.042253  0.793444
           XYZ     0.659067  1.967308 -2.562109 -0.874071  0.416899

[10093 rows x 5 columns]

In [80]:
betas.index

MultiIndex([('2020-08-31', 'AAPL'),
            ('2020-08-31', 'ABBV'),
            ('2020-08-31',  'ABT'),
            ('2020-08-31',  'ACN'),
            ('2020-08-31', 'ADBE'),
            ('2020-08-31',  'ADP'),
            ('2020-08-31', 'ADSK'),
            ('2020-08-31', 'ALGN'),
            ('2020-08-31', 'AMAT'),
            ('2020-08-31',  'AMD'),
            ...
            ('2026-06-30',  'UPS'),
            ('2026-06-30',    'V'),
            ('2026-06-30', 'VRTX'),
            ('2026-06-30',   'VZ'),
            ('2026-06-30', 'WDAY'),
            ('2026-06-30',  'WDC'),
            ('2026-06-30',  'WFC'),
            ('2026-06-30',  'WMT'),
            ('2026-06-30',  'XOM'),
            ('2026-06-30',  'XYZ')],
           names=['date', 'ticker'], length=10093)

In [81]:
factors = ['Mkt-RF', 'SMB', 'HML', 'RMW', 'CMA']

data = (data.join(betas.groupby('ticker').shift()))

Yeh .shift() isliye karna pada kyuki manlo august month ka jo data woh data toh kam sepetember month ke analysis ke liye kam aayega na toh actually august ka kam september ke decisions ko shape karta toh august ka data september ke liye meaningful hai na ki august ke liye isliye ye chiz model ko smjhane ke liye hume data ko ek position aage shift karna pada tha warna agr nhi karenge toh bhut complex ho jayega ml model ke liye smjhna  

In [83]:
data.loc[:, factors] = data.groupby('ticker', group_keys=False)[factors].apply(lambda x: x.fillna(x.mean()))

In [84]:
data

close        high  garman_klass_vol        rsi  \
date       ticker                                                        
2020-08-31 AAPL    125.057526  126.957042          0.000707  76.744809   
           ABBV     75.952614   76.142956          0.000105  52.776389   
           ABT      97.865204   99.894561          0.000204  67.784280   
           ACN     217.961929  221.241397          0.000094  67.876178   
           ADBE    513.390015  516.729980          0.000153  66.121887   
...                       ...         ...               ...        ...   
2026-08-31 WDC     454.100006  467.799988          0.000662  43.796277   
           WFC      88.930000   89.220001          0.000170  59.174628   
           WMT     116.010002  116.050003          0.000153  54.108979   
           XOM     158.722244  159.199152          0.000119  62.142821   
           XYZ      78.279999   79.199997          0.000312  49.935831   

                     bb_low    bb_mid   bb_high       atr      macd  \
date       ticker                                                     
2020-08-31 AAPL    4.634849  4.743995  4.853141 -0.215957  1.968695   
           ABBV    4.304350  4.329241  4.354133 -1.060942 -0.352670   
           ABT     4.460974  4.528430  4.595886  0.635079  1.529678   
           ACN     5.324580  5.364798  5.405016 -1.217038  0.870915   
           ADBE    6.036542  6.152162  6.267782  0.679071  1.881340   
...                     ...       ...       ...       ...       ...   
2026-08-31 WDC     6.028525  6.205588  6.382652  5.179602 -4.059220   
           WFC     4.444806  4.474222  4.503639  1.216887  0.744505   
           WMT     4.693527  4.727599  4.761672  2.212910 -0.599633   
           XOM     4.988154  5.038804  5.089454  2.227161  1.943697   
           XYZ     4.342263  4.398908  4.455553 -0.668036  0.144249   

                   return_1m  return_2m  return_3m  return_6m  return_9m  \
date       ticker                                                          
2020-08-31 AAPL     0.204976   0.176611   0.158586   0.110026   0.076700   
           ABBV     0.009061  -0.006341   0.015125   0.023252   0.014328   
           ABT      0.087738   0.096338   0.050042   0.059706   0.029314   
           ACN      0.067399   0.059010   0.060994   0.049919   0.021139   
           ADBE     0.155451   0.085986   0.099162   0.068430   0.057830   
...                      ...        ...        ...        ...        ...   
2026-08-31 WDC     -0.166544  -0.156820  -0.050854   0.084251   0.281101   
           WFC      0.034593   0.040332   0.048744   0.016685   0.005747   
           WMT      0.043255   0.012067   0.000748  -0.015546   0.006081   
           XOM      0.027728   0.080947   0.032203   0.008908   0.037822   
           XYZ     -0.036435   0.014889   0.011145   0.034948   0.017777   

                   return_12m    Mkt-RF       SMB       HML       RMW  \
date       ticker                                                       
2020-08-31 AAPL      0.073806  1.018473  0.091631 -0.410227  0.442392   
           ABBV      0.036576  0.488717 -0.315094  0.399168  0.138334   
           ABT       0.022377  0.797525 -0.230780  0.080680 -0.126645   
           ACN       0.017444  1.055670 -0.080026 -0.175076 -0.012285   
           ADBE      0.050419  1.594977 -0.804373  0.044982  0.310262   
...                       ...       ...       ...       ...       ...   
2026-08-31 WDC       0.195755       NaN       NaN       NaN       NaN   
           WFC       0.008411  1.025111 -0.081648  1.304222 -0.959372   
           WMT       0.015550  0.712885 -0.102831 -0.285567  0.416614   
           XOM       0.030196  0.677901 -0.234842  0.975004  0.033052   
           XYZ      -0.001434  2.245977  0.630212 -0.304450 -2.658683   

                        CMA  
date       ticker            
2020-08-31 AAPL   -0.082594  
           ABBV    0.707258  
           ABT     0.747452  
           ACN    -0.253985  
           ADBE   

In [85]:
data = data.drop('close', axis=1)

In [86]:
data = data.dropna()

In [87]:
data.info()

<class 'pandas.DataFrame'>
MultiIndex: 10135 entries, (Timestamp('2020-08-31 00:00:00'), 'AAPL') to (Timestamp('2026-08-31 00:00:00'), 'XYZ')
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   high              10135 non-null  float64
 1   garman_klass_vol  10135 non-null  float64
 2   rsi               10135 non-null  float64
 3   bb_low            10135 non-null  float64
 4   bb_mid            10135 non-null  float64
 5   bb_high           10135 non-null  float64
 6   atr               10135 non-null  float64
 7   macd              10135 non-null  float64
 8   return_1m         10135 non-null  float64
 9   return_2m         10135 non-null  float64
 10  return_3m         10135 non-null  float64
 11  return_6m         10135 non-null  float64
 12  return_9m         10135 non-null  float64
 13  return_12m        10135 non-null  float64
 14  Mkt-RF            10135 non-null  float64
 15  SMB               101

In [89]:
data.to_pickle('data/04_famafrench_stocksWithBetas.pkl')